File: examples/hello_service/hello_service_py/hello_service_py/hello_client.py
#!/usr/bin/env python3

'''
    A minimal ROS2 python node that subscribes to a service.

    Basic structure for ROS2 node publisher:
    - Initialise rclpy
    - Create a node
    - Create the client for the service
    - Keep it alive with spin
    - Shutdown

    RELIABLE and VOLATILE have been selected for the QOS_PROFILE to match with
    the ones defined in the service server.

    Author: Ioannis Selinis <selinis.g@gmail.com> 2025
'''

import random
from functools import partial  # partial to pass args in add_done_callback
import rclpy
from rclpy.node import Node
from rclpy.task import Future
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from example_interfaces.srv import AddTwoInts


# https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html
QOS_PROFILE = QoSProfile(
    reliability=ReliabilityPolicy.RELIABLE,         # RELIABLE or BEST_EFFORT
    durability=DurabilityPolicy.VOLATILE,           # VOLATILE or TRANSIENT_LOCAL
    history=HistoryPolicy.KEEP_LAST,                # KEEP_ALL or KEEP_LAST
    depth=1                                         # Used only with KEEP_LAST
)


class HelloClient(Node):
    """The HelloClient class that inherits from rclpy.node.Node.
    Initialises with the client for the service AddTwoInts.
    """
    def __init__(self):
        super().__init__(node_name='hello_client')
        self.client = self.create_client(srv_type=AddTwoInts,
                                         srv_name='add_two_ints',
                                         qos_profile=QOS_PROFILE)
        # Calling a non existing service doesn't throw an error,
        # but it's good to check if it's up or not.
        while not self.client.wait_for_service(1.0):
            self.get_logger().warning('Waiting for the service add_two_ints...')
        self.timer = self.create_timer(timer_period_sec=1.0,
                                       callback=self.call_service)


    def call_service(self) -> None:
        """Triggered every second and calls the service"""
        request = AddTwoInts.Request()
        request.a = random.randint(5, 35)
        request.b = random.randint(2, 102)
        future = self.client.call_async(request=request)
        future.add_done_callback(partial(self.service_callback, request=request))


    def service_callback(self, future: Future, request: AddTwoInts.Request) -> None:
        """Callback for the service call"""
        self.get_logger().info(f'Result: {future.result()} for {request.a} + {request.b}')


def main(args=None) -> None:
    """Entry Point for our node"""
    rclpy.init(args=args)   # Initialise rclpy (ROS communications)
    node = HelloClient()
    rclpy.spin(node=node)   # Keep node alive
    node.destroy_node()     # Destroy node
    rclpy.shutdown()        # Shutdown the executor


if __name__ == '__main__':
    main()

File: examples/hello_service/hello_service_cpp/src/hello_client.cpp
/**
 *  A minimal ROS2 cpp node that creates a service.
 *
 *  Basic structure for ROS2 node:
 *  - Initialise rclcpp
 *  - Create a node
 *  - Create the client for the service
 *  - Keep it alive with spin
 *  - Shutdown
 *
 *  RELIABLE and VOLATILE have been selected for the QOS_PROFILE to match with those
 *  defined in the service server.
 *
 *  Author: Ioannis Selinis <selinis.g@gmail.com> 2025
 */

#include <random>
#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp>


static const rmw_qos_profile_t rmw_qos_profile =
{
    RMW_QOS_POLICY_HISTORY_KEEP_LAST,
    1,
    RMW_QOS_POLICY_RELIABILITY_RELIABLE,
    RMW_QOS_POLICY_DURABILITY_VOLATILE,
    RMW_QOS_DEADLINE_DEFAULT,
    RMW_QOS_LIFESPAN_DEFAULT,
    RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
    RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
    false
};


class HelloClient : public rclcpp::Node
{
    public:
        HelloClient() : Node("hello_c_client")
        {
            m_rng = std::mt19937 (m_rd());
            m_distr = std::uniform_int_distribution<int>(0, 237);
            m_client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints", rmw_qos_profile);
            while (!m_client->wait_for_service(std::chrono::seconds(1))){
                RCLCPP_WARN(this->get_logger(), "Waiting for the server...");
            }
            m_timer = this->create_wall_timer(std::chrono::seconds(1),
                                              std::bind(&HelloClient::callService, this));
            RCLCPP_INFO(this->get_logger(), "Node is ready");
        }

    private:
        void callService()
        {
            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            request->a = m_distr(m_rng);
            request->b = m_distr(m_rng);
            m_client->async_send_request(request,
                                         std::bind(&HelloClient::responseService, this, std::placeholders::_1));
        }
        void responseService(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future)
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "Response sum: %d", (int) response->sum);
        }
        rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr m_client;
        rclcpp::TimerBase::SharedPtr m_timer;
        std::random_device m_rd;
        std::mt19937 m_rng;
        std::uniform_int_distribution<int> m_distr;
};


int main (int argc, char **argv)
{
    rclcpp::init(argc, argv);                       // Initialise rclcpp (ROS communications)
    auto node = std::make_shared<HelloClient>();    // Smart/Shared pointer
    rclcpp::spin(node);                             // Keep node alive
    rclcpp::shutdown();                             // Shutdown executor
}
